modules/cmd.js

```diff
@@ -0,0 +1,150 @@
+import { throwInvariant, flatten, promisify } from './utils';
+
+
+const isCmdSymbol = Symbol('isCmd');
+
+
+const cmdTypes = {
+  PROMISE: 'PROMISE',
+  CALL: 'CALL',
+  CALLBACK: 'CALLBACK',
+  CONSTANT: 'CONSTANT',
+  BATCH: 'BATCH',
+  MAP: 'MAP',
+  NONE: 'NONE',
+}
+
+
+export const isCmd = (object) => {
+  return object ? object[isCmdSymbol] : false
+}
+
+
+export const cmdToPromise = (cmd) => {
+  switch (cmd.type) {
+    case cmdTypes.PROMISE:
+      return cmd.promiseFactory(...cmd.args)
+        .then(
+          cmd.successActionCreator,
+          cmd.failureActionCreator
+        )
+        .then((action) => [action])
+
+    case cmdTypes.CALL:
+      const result = cmd.resultFactory(...cmd.args)
+      return Promise.resolve([cmd.actionCreator(result)])
+
+    case cmdTypes.CALLBACK:
+      return promisify(cmd.nodeStyleFunction)(...cmd.args)
+        .then(
+          cmd.successActionCreator,
+          cmd.failureActionCreator
+        )
+        .then((action) => [action])
+
+    case cmdTypes.CONSTANT:
+      return Promise.resolve(cmd.action)
+
+    case cmdTypes.BATCH:
+      const filteredCmds = cmd.cmds
+        .filter((nextCmd) => nextCmd.type !== cmdTypes.NONE)
+
+      const batchedPromises = filteredCmds.map(cmdToPromise).filter((x) => x)
```

Comment:
yeah i'll get rid of the first filter. nested empty batches will also return `null` so both cases need to be handled.